File: java/src/main/java/dependencies/Project.java
package dependencies;

import static dependencies.TestStatus.*;

public class Project {
    private final boolean buildsSuccessfully;
    private final TestStatus testStatus;

    public static ProjectBuilder builder() {
        return new ProjectBuilder();
    }

    private Project(boolean buildsSuccessfully, TestStatus testStatus) {
        this.buildsSuccessfully = buildsSuccessfully;
        this.testStatus = testStatus;
    }

    public boolean hasTests() {
        return testStatus != NO_TESTS;
    }

    public String runTests() {
        return testStatus == PASSING_TESTS ? "success" : "failure";
    }

    public String deploy() {
        return buildsSuccessfully ? "success" : "failure";
    }

    public static class ProjectBuilder {
        private boolean buildsSuccessfully;
        private TestStatus testStatus;

        public ProjectBuilder setTestStatus(TestStatus testStatus) {
            this.testStatus = testStatus;
            return this;
        }

        public ProjectBuilder setDeploysSuccessfully(boolean buildsSuccessfully) {
            this.buildsSuccessfully = buildsSuccessfully;
            return this;
        }

        public Project build() {
            return new Project(buildsSuccessfully, testStatus);
        }
    }
}

File: typescript/src/dependencies/Project.ts
import {TestStatus} from "./TestStatus";

export class Project {
    private readonly buildsSuccessfully: boolean;
    private readonly testStatus: TestStatus;

    static builder(): ProjectBuilder {
        return new ProjectBuilder();
    }

    constructor(buildsSuccessfully: boolean, testStatus: TestStatus) {
        this.buildsSuccessfully = buildsSuccessfully;
        this.testStatus = testStatus;
    }

    hasTests(): boolean {
        return this.testStatus != TestStatus.NO_TESTS;
    }

    runTests(): string {
        return this.testStatus == TestStatus.PASSING_TESTS ? "success" : "failure";
    }


    deploy(): string {
        return this.buildsSuccessfully ? "success" : "failure";
    }
}

export class ProjectBuilder {
    private buildsSuccessfully: boolean;
    private testStatus: TestStatus;

    setTestStatus(testStatus: TestStatus): ProjectBuilder {
        this.testStatus = testStatus;
        return this;
    }

    setDeploysSuccessfully(buildsSuccessfully: boolean): ProjectBuilder {
        this.buildsSuccessfully = buildsSuccessfully;
        return this;
    }

    build(): Project {
        return new Project(this.buildsSuccessfully, this.testStatus);
    }
}
